File: src/ShowyEdge/objc/WorkSpaceData.m
@import Carbon;
#import "WorkSpaceData.h"
#import "NotificationKeys.h"
#import "weakify.h"

@interface WorkSpaceData ()

@property(copy, readwrite) NSString* currentInputSourceID;
@property(copy, readwrite) NSString* currentInputModeID;
@property(copy, readwrite) NSSet* menubarOrigins;

@end

@implementation WorkSpaceData

- (void)observer_kTISNotifySelectedKeyboardInputSourceChanged:(NSNotification*)notification {
  dispatch_async(dispatch_get_main_queue(), ^{
    TISInputSourceRef ref = TISCopyCurrentKeyboardInputSource();
    if (ref) {
      NSString* currentInputSourceID = (__bridge NSString*)(TISGetInputSourceProperty(ref, kTISPropertyInputSourceID));
      NSString* currentInputModeID = (__bridge NSString*)(TISGetInputSourceProperty(ref, kTISPropertyInputModeID));

      if (currentInputSourceID.length == 0) {
        currentInputSourceID = @"unknown";
      }
      if (currentInputModeID.length == 0) {
        currentInputModeID = @"";
      }

      self.currentInputSourceID = currentInputSourceID;
      self.currentInputModeID = currentInputModeID;

      [NSNotificationCenter.defaultCenter postNotificationName:kCurrentInputSourceIDChangedNotification object:nil];

      CFRelease(ref);
    }
  });
}

- (instancetype)init {
  self = [super init];

  if (self) {
    self.currentInputSourceID = @"";
    self.currentInputModeID = @"";
    self.menubarOrigins = NSSet.new;

    // In Mac OS X 10.7, NSDistributedNotificationCenter is suspended after calling [NSAlert runModal].
    // So, we need to set suspendedDeliveryBehavior to NSNotificationSuspensionBehaviorDeliverImmediately.
    [NSDistributedNotificationCenter.defaultCenter addObserver:self
                                                      selector:@selector(observer_kTISNotifySelectedKeyboardInputSourceChanged:)
                                                          name:(NSString*)(kTISNotifySelectedKeyboardInputSourceChanged)
                                                          object:nil
                                            suspensionBehavior:NSNotificationSuspensionBehaviorDeliverImmediately];

    @weakify(self);
    [NSWorkspace.sharedWorkspace.notificationCenter addObserverForName:NSWorkspaceActiveSpaceDidChangeNotification
                                                                object:nil
                                                                 queue:NSOperationQueue.mainQueue
                                                            usingBlock:^(NSNotification* note) {
                                                              @strongify(self);
                                                              if (!self) return;

                                                              [self updateMenubarOrigins];
                                                            }];
  }

  return self;
}

- (void)setup {
  [self observer_kTISNotifySelectedKeyboardInputSourceChanged:nil];
  [self updateMenubarOrigins];
}

- (void)dealloc {
  [NSDistributedNotificationCenter.defaultCenter removeObserver:self];
}

- (void)updateMenubarOrigins {
  NSArray* windows = CFBridgingRelease(CGWindowListCopyWindowInfo(kCGWindowListOptionOnScreenOnly, kCGNullWindowID));
  NSMutableSet* menubarOrigins = NSMutableSet.new;

  // We detect full screen spaces by checking if there's a menubar in the window list.
  // If not, we assume it's in fullscreen mode.
  for (NSDictionary* d in windows) {
    if ([d[@"kCGWindowOwnerName"] isEqualToString:@"Window Server"] &&
        [d[@"kCGWindowName"] isEqualToString:@"Menubar"]) {
      NSDictionary* bounds = d[@"kCGWindowBounds"];
      if (bounds) {
        NSNumber* x = bounds[@"X"];
        NSNumber* y = bounds[@"Y"];
        [menubarOrigins addObject:@{
          @"x" : x,
          @"y" : y,
        }];
      }
    }
  }

  if ([self.menubarOrigins isEqualToSet:menubarOrigins]) {
    return;
  }

  self.menubarOrigins = menubarOrigins;

  [NSNotificationCenter.defaultCenter postNotificationName:kFullScreenModeChangedNotification object:nil];
}

@end

File: src/ShowyEdge/swift/WorkspaceData.swift
import AppKit
import Carbon
import CoreGraphics
import Foundation

@objc
public class WorkspaceData: NSObject {
  @objc
  public static let shared = WorkspaceData()
  @objc
  public static let fullScreenModeChanged = Notification.Name("fullScreenModeChanged")
  @objc
  public static let currentInputSourceChanged = Notification.Name("currentInputSourceChanged")

  var currentInputSourceID: String = ""
  var currentInputModeID: String = ""
  var menubarOrigins: Set<CGPoint> = []
  var activeSpaceDidChangeObserver: NSObjectProtocol?

  @objc
  public func start() {
    stop()

    DistributedNotificationCenter.default().addObserver(
      self,
      selector: #selector(selectedKeyboardInputSourceChanged),
      name: kTISNotifySelectedKeyboardInputSourceChanged as NSNotification.Name?,
      object: nil,
      suspensionBehavior: .deliverImmediately
    )

    activeSpaceDidChangeObserver = NSWorkspace.shared.notificationCenter.addObserver(
      forName: NSWorkspace.activeSpaceDidChangeNotification,
      object: nil,
      queue: OperationQueue.main,
      using: { [weak self] _ in
        guard let self = self else { return }

        self.updateMenubarOrigins()
      }
    )

    updateMenubarOrigins()
  }

  @objc
  public func stop() {
    DistributedNotificationCenter.default.removeObserver(self)

    if activeSpaceDidChangeObserver != nil {
      NSWorkspace.shared.notificationCenter.removeObserver(activeSpaceDidChangeObserver!)
    }
  }

  @objc
  private func selectedKeyboardInputSourceChanged() {
    DispatchQueue.main.async {
      let inputSource = TISCopyCurrentKeyboardInputSource().takeRetainedValue()

      self.currentInputSourceID = inputSource.inputSourceID ?? "unknown"
      self.currentInputModeID = inputSource.inputModeID ?? ""

      NotificationCenter.default.post(name: WorkspaceData.currentInputSourceChanged,
                                      object: nil)
    }
  }

  private func updateMenubarOrigins() {
    var menubarOrigins: Set<CGPoint> = []

    if let windows = CGWindowListCopyWindowInfo(.optionOnScreenOnly, kCGNullWindowID) as? [[String: Any]] {
      // We detect full screen spaces by checking if there's a menubar in the window list.
      // If not, we assume it's in fullscreen mode.
      for dict in windows {
        if dict["kCGWindowOwnerName"] as? String == "Window Server",
           dict["kCGWindowName"] as? String == "Menubar"
        {
          if let bounds = dict["kCGWindowBounds"] as? [String: Any],
             let x = bounds["X"] as? NSNumber,
             let y = bounds["Y"] as? NSNumber
          {
            menubarOrigins.insert(CGPoint(x: x.doubleValue,
                                          y: y.doubleValue))
          }
        }
      }

      if self.menubarOrigins != menubarOrigins {
        self.menubarOrigins = menubarOrigins

        NotificationCenter.default.post(name: WorkspaceData.fullScreenModeChanged,
                                        object: nil)
      }
    }
  }
}
